### PR TITLE
MES-3798: Using category specific tests reducer

### PR DESCRIPTION
**Pull request by szabikr**
## Description

Using category specific reducers where needed.
Reducer factory implementations

## Checklist

- [x] PR title includes the JIRA ticket number
- [x] Branch is rebased against the latest develop
- [x] Code has been tested manually
- [x] PR link added to JIRA ticket
- [x] One review from each scrum team
- [x] Squashed commit contains the JIRA ticket number

### CODE REVIEW

File: src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import { PassCompletion } from '@dvsa/mes-test-schema/categories/BE';
+import { Action } from '@ngrx/store';
+
+export const initialState: PassCompletion = {
+  provisionalLicenceProvided: false,
+  passCertificateNumber: '',
+  code78Present: false,
+};
+
+export function passCompletionCatBEReducer(state: PassCompletion = initialState, action: Action) {
```
Comment (TomBillingtonUK) [inline src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts]: missing return type and should be Cat Be not Cat BE

File: src/modules/tests/test-data/manoeuvres/manoeuvres.cat-be.reducer.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { Manoeuvres } from '@dvsa/mes-test-schema/categories/BE';
+import { Action } from '@ngrx/store';
+
+export const initialState: Manoeuvres = {
+  reverseLeft: {},
+};
+
+export function manoeuvresCatBEReducer(state: Manoeuvres = initialState, action: Action) {
```
Comment (TomBillingtonUK) [inline src/modules/tests/test-data/manoeuvres/manoeuvres.cat-be.reducer.ts]: missing return type and should be Cat Be not Cat BE

File: src/modules/tests/test-data/test-data.cat-be.reducer.ts [hunk 3]
```diff
@@ -1,31 +1,29 @@
...
 import { eyesightTestReducer } from './eyesight-test/eyesight-test.reducer';
-import { manoeuvresReducer } from './manoeuvres/manoeuvres.reducer';
+import { manoeuvresCatBEReducer } from './manoeuvres/manoeuvres.cat-be.reducer';
 import { seriousFaultsReducer } from './serious-faults/serious-faults.reducer';
 import { testRequirementsReducer } from './test-requirements/test-requirements.reducer';
-import { vehicleChecksReducer } from './vehicle-checks/vehicle-checks.reducer';
+import { vehicleChecksCatBEReducer } from './vehicle-checks/vehicle-checks.cat-be.reducer';
 
-export function testDataCatBeReducer(
+export function testDataCatBEReducer(
```
Comment (TomBillingtonUK) [inline src/modules/tests/test-data/test-data.cat-be.reducer.ts]: should be Be not BE

File: src/modules/tests/tests-reducer-factory.ts [hunk 4]
```diff
@@ -2,14 +2,19 @@
 import { TestCategory } from '../../shared/models/test-category';
 import { Action } from '@ngrx/store';
 import { StandardCarTestCATBSchema } from '@dvsa/mes-test-schema/categories/B';
+import { StandardTrailerTestCATBESchema } from '@dvsa/mes-test-schema/categories/BE';
 import { testsCatBReducer } from './tests.cat-b.reducer';
+import { testsCatBEReducer } from './tests.cat-be.reducer';
+import { TestResultUnionType } from './tests.model';
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests-reducer-factory.ts:8]: Should use union in api definitions

File: src/modules/tests/tests.cat-be.reducer.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+import { rekeyReasonReducer } from './rekey-reason/rekey-reason.reducer';
+import { examinerBookedReducer } from './examiner-booked/examiner-booked.reducer';
+import { examinerConductedReducer } from './examiner-conducted/examiner-conducted.reducer';
+import { examinerKeyedReducer } from './examiner-keyed/examiner-keyed.reducer';
+import { changeMarkerReducer } from './change-marker/change-marker';
+import { activityCodeReducer } from './activity-code/activity-code.reducer';
+import { journalDataReducer } from './journal-data/journal-data.reducer';
+import { testDataCatBEReducer } from './test-data/test-data.cat-be.reducer';
+
+export function testsCatBEReducer(
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.cat-be.reducer.ts]: should be Be

File: src/modules/tests/tests.model.ts [hunk 6]
```diff
@@ -1,12 +1,15 @@
 import { StandardCarTestCATBSchema } from '@dvsa/mes-test-schema/categories/B';
 import { TestStatus } from './test-status/test-status.model';
+import { StandardTrailerTestCATBESchema } from '@dvsa/mes-test-schema/categories/BE';
+
+export type TestResultUnionType = StandardCarTestCATBSchema | StandardTrailerTestCATBESchema;
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.model.ts:5]: can be removed if we use the one in api definitions

File: src/modules/tests/tests.reducer.ts [hunk 7]
```diff
@@ -64,13 +65,12 @@ const deriveCategory = (state: TestsModel, action: Action, slotId: string | null
...
 
-const createStateObject = (state: TestsModel, action: Action, slotId: string, category: string) => {
+const createStateObject = (state: TestsModel, action: Action, slotId: string, category: string): TestsModel => {
   return {
     ...state,
     startedTests: {
       ...state.startedTests,
       [slotId]: {
-        ...state.startedTests[slotId],
         // TODO - Need to get category from somehwere else as at start of test it's null
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.reducer.ts]: remove TODO

File: src/modules/tests/tests.selector.ts [hunk 8]
```diff
@@ -10,7 +10,7 @@ import { DateTime } from '../../shared/helpers/date-time';
 
 export const getCurrentTestSlotId = (tests: TestsModel): string => tests.currentTest.slotId;
 
-export const getCurrentTest = (tests: TestsModel): StandardCarTestCATBSchema => {
+export const getCurrentTest = (tests: TestsModel): TestResultUnionType => {
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.selector.ts:13]: Wrong union

File: src/modules/tests/tests.selector.ts [hunk 9]
```diff
@@ -20,7 +20,7 @@ export const getCurrentTestStatus = (tests: TestsModel): TestStatus => {
   return tests.testStatus[currentTestSlotId];
 };
 
-export const getTestById = (tests: TestsModel, slotId: string): StandardCarTestCATBSchema => {
+export const getTestById = (tests: TestsModel, slotId: string): TestResultUnionType => {
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.selector.ts:23]: Wrong union

File: src/modules/tests/tests.selector.ts [hunk 10]
```diff
@@ -54,7 +54,7 @@ export const isTestOutcomeSet = (test: StandardCarTestCATBSchema) => {
   return false;
 };
 
-export const isPassed = (test: StandardCarTestCATBSchema): boolean => {
+export const isPassed = (test: TestResultUnionType): boolean => {
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.selector.ts:57]: Wrong union

File: src/modules/tests/tests.selector.ts [hunk 11]
```diff
@@ -93,13 +93,13 @@ export const getIncompleteTestsSlotIds = (tests: TestsModel): string[] => {
     && tests.testStatus[slotId] !== TestStatus.Completed);
 };
 
-const isTestBeforeToday = (test: StandardCarTestCATBSchema): boolean => {
+const isTestBeforeToday = (test: TestResultUnionType): boolean => {
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.selector.ts:96]: Wrong union

File: src/modules/tests/tests.selector.ts [hunk 12]
```diff
@@ -93,13 +93,13 @@ export const getIncompleteTestsSlotIds = (tests: TestsModel): string[] => {
...
   const testDate = new DateTime(test.journalData.testSlotAttributes.start);
   const today = new DateTime();
   return today.daysDiff(new Date(testDate.format('YYYY-MM-DD'))) < 0;
 };
 
-export const getIncompleteTests = (tests: TestsModel): StandardCarTestCATBSchema[] => {
+export const getIncompleteTests = (tests: TestsModel): TestResultUnionType[] => {
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.selector.ts:102]: Wrong union

File: src/modules/tests/tests.selector.ts [hunk 13]
```diff
@@ -109,10 +109,10 @@ export const getIncompleteTestsCount = (tests: TestsModel): number => {
   return incompleteTestsSlotIds.length;
 };
 
-export const getOldestIncompleteTest = (tests: TestsModel): StandardCarTestCATBSchema => {
+export const getOldestIncompleteTest = (tests: TestsModel): TestResultUnionType => {
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.selector.ts:112]: Wrong union

File: src/modules/tests/tests.selector.ts [hunk 14]
```diff
@@ -109,10 +109,10 @@ export const getIncompleteTestsCount = (tests: TestsModel): number => {
...
   const incompleteTestsSlotIds: string[] = getIncompleteTestsSlotIds(tests);
 
-  let oldestTest: StandardCarTestCATBSchema;
+  let oldestTest: TestResultUnionType;
```
Comment (TomBillingtonUK) [inline src/modules/tests/tests.selector.ts:115]: Wrong union

File: src/providers/test-submission/test-submission.ts [hunk 15]
```diff
@@ -15,11 +14,12 @@ import { LogType } from '../../shared/models/log.model';
 import { LogHelper } from '../logs/logsHelper';
 import { AppConfigProvider } from '../app-config/app-config';
 import { TestStatus } from '../../modules/tests/test-status/test-status.model';
+import { TestResultUnionType } from '../../modules/tests/tests.model';
 
 export interface TestToSubmit {
   index: number;
   slotId: string;
-  payload: StandardCarTestCATBSchema;
+  payload: TestResultUnionType;
```
Comment (TomBillingtonUK) [inline src/providers/test-submission/test-submission.ts:22]: Wrong union

Comment (TomBillingtonUK) [reply src/providers/test-submission/test-submission.ts:22]: Same throughout this file - i'm getting lazy pointing them all out

**Review by TomBillingtonUK — COMMENTED** [review]
Looks good - lots of comments

File: src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts [hunk 16]
```diff
@@ -0,0 +1,16 @@
+
+import { PassCompletion } from '@dvsa/mes-test-schema/categories/BE';
+import { Action } from '@ngrx/store';
+
+export const initialState: PassCompletion = {
+  provisionalLicenceProvided: false,
```
Comment (TomBillingtonUK) [inline src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts]: the data values need to be null else you are submitting data that may fail validation and/or isn't needed for a particular test type

File: src/modules/tests/tests.model.ts (hunk 6 above)
Comment (ammarhaiderbjss) [reply src/modules/tests/tests.model.ts:5]: I think we'll merge this work in as-is and then do the work to implement the new real schemas later

File: src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts (hunk 1 above)
Comment (szabikr) [reply src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts]: Changed it

File: src/modules/tests/test-data/manoeuvres/manoeuvres.cat-be.reducer.ts (hunk 2 above)
Comment (szabikr) [reply src/modules/tests/test-data/manoeuvres/manoeuvres.cat-be.reducer.ts]: added it, changed it

File: src/modules/tests/test-data/test-data.cat-be.reducer.ts (hunk 3 above)
Comment (szabikr) [reply src/modules/tests/test-data/test-data.cat-be.reducer.ts]: changed it

File: src/modules/tests/tests-reducer-factory.ts (hunk 4 above)
Comment (szabikr) [reply src/modules/tests/tests-reducer-factory.ts:8]: Will add it as the next PR when we use all the new schemas

File: src/modules/tests/tests.cat-be.reducer.ts (hunk 5 above)
Comment (szabikr) [reply src/modules/tests/tests.cat-be.reducer.ts]: changed it

File: src/modules/tests/tests.reducer.ts (hunk 7 above)
Comment (szabikr) [reply src/modules/tests/tests.reducer.ts]: removed it

File: src/modules/tests/tests.selector.ts (hunk 8 above)
Comment (szabikr) [reply src/modules/tests/tests.selector.ts:13]: Will add it as part of next PR

File: src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts (hunk 16 above)
Comment (szabikr) [reply src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts]: Okay, changing it

Comment (szabikr) [reply src/modules/tests/pass-completion/pass-completion.cat-be.reducer.ts]: done

**Review by ammarhaiderbjss — APPROVED** [review]

**Review by MarkPatt89 — APPROVED** [review]
